treeherder/webapp/api/bug_creation.py

```diff
@@ -0,0 +1,66 @@
+import logging
+
+from rest_framework import viewsets
+from rest_framework.decorators import action
+from rest_framework.response import Response
+
+from treeherder.model.models import FilesBugzillaMap
+from treeherder.webapp.api import serializers
+
+logger = logging.getLogger(__name__)
+
+
+class FilesBugzillaMapViewSet(viewsets.ReadOnlyModelViewSet):
+    @action(detail=True, methods=['get'])
+    def get_queryset(self):
+        """
+        Gets a set of bug suggestions for this job
+        """
+        path = self.request.query_params.get('path')
+        path = path.replace('\\', '/')
+        # Drop parameters
+        path = (path.split('?'))[0]
+        file = (path.split('/'))[-1]
+        fileNameParts = file.split('.')
+        file_without_extension = fileNameParts[0] + ('.' if len(fileNameParts) > 1 else '')
+        queryset = (
+            FilesBugzillaMap.objects.select_related('bugzilla_component')
+            .filter(path__endswith=path)
+            .exclude(path__startswith='testing/web-platform/meta/')
+            .values('bugzilla_component__product', 'bugzilla_component__component')
+            .distinct()
+        )
+        if len(queryset) == 0:
+            # E.g. web-platform-tests ("wpt") can use test files generated from
+            # other files which just have different file extensions.
+            path_without_extension = (path.rsplit('/', 1))[0] + '/' + file_without_extension
+            queryset = (
+                FilesBugzillaMap.objects.select_related('bugzilla_component')
+                .filter(path__contains=path_without_extension)
+                .exclude(path__startswith='testing/web-platform/meta/')
+                .values('bugzilla_component__product', 'bugzilla_component__component')
+                .distinct()
+            )
+        if len(queryset) > 0:
+            return queryset[: min(5, len(queryset))]
+        queryset = (
+            FilesBugzillaMap.objects.select_related('bugzilla_component')
+            .filter(file_name=file)
+            .values('bugzilla_component__product', 'bugzilla_component__component')
+            .distinct()
```

Comment:
This is a lot of queries... it seems like it could be slow when you get a certain number of rows in the table. Since you're essentially repeating the same query several times you could create a new queryset class in models.py and add it as a 'manager' to the FilesBugzillaMap. You'll see an example in the top of the models.py file. 

And/or you could use Q objects to filter with 'OR' logic in the queryset.

Reply:
For context: This feature gets used 20-50 times per day, almost exclusively by code sheriffs.
* The tests which are not web-platform-tests will almost all be covered by the first query.
* The second query is more an edge case.
* The big share of web-platform-test failures will have results after query 3.
* The web-platform-tests which use a generated file name need query #4.
The `path` and `file_name` columns are indexed.

Regarding the use of a custom manager:
The rule `.filter(file_name__startswith=file_without_extension)` from the last query could be used to define the base queryset for the manager but the other filtering rules of query 1-3 would still need to be applied in custom methods. models.py's `FailuresQuerySet` executes the query each time from the full row set (by calling `self.<method>`) if I understand it correctly.

A request to the API endpoint is rather quick based on local testing, e.g. faster than the generation of the bugs suggested for failure lines if there are several failure lines, even with empty `bugscache`.

Can we land the current code for the bugzilla product+component endpoint like it is and evaluate if further work is needed based on performance in production? If the code should be written or organized different, could you provide further details, please?

Reply:
> The rule .filter(file_name__startswith=file_without_extension) from the last query could be used to define the base queryset for the manager but the other filtering rules of query 1-3 would still need to be applied in custom methods. models.py's FailuresQuerySet executes the query each time from the full row set (by calling self.<method>) if I understand it correctly

I was trying to figure out how to make the code a little more DRY, but maybe its not worth the effort. Do you feel confident trying this out on stage or do you want to test it on prototype branch first?

Reply:
Having it run on prototype and checking the logs the import succeeds there would be beneficial. After the recent docker upgrade, 2 imports had failed locally because the docker container crashed. Memory was set to 2 GB.

Reply:
Ok, so you'll want to push your branch to the prototype branch in this repo and your pr should automatically deploy to the prototype instance. Then if you file a bug with cloudOps, they can add set up the cron job. Since oremj is leaving it might not be very quick... no idea who is replacing him as our tech support.